src/splice.ts

```diff
@@ -1,12 +1,133 @@
+import type {
+  ArraySplice,
+  GreaterThan,
+  IfNever,
+  IsLiteral,
+  IsNegative,
+  Subtract,
+  Sum,
+} from "type-fest";
+import type {
+  CoercedArray,
+  IterableContainer,
+  TupleParts,
+  WidenLiteral,
+} from "./internal/types";
 import { purry } from "./purry";
 
+type IfLiteral<N extends number, Fallback> = number extends N ? Fallback : N;
+
+type PositiveIndex<
+  T extends IterableContainer,
+  N extends number,
+> = N extends unknown
+  ? IsNegative<N> extends false
+    ? GreaterThan<N, T["length"]> extends true
+      ? T["length"]
+      : N
+    : number extends T["length"]
+      ? number
+      : IfLiteral<Sum<T["length"], N>, 0>
+  : never;
+
+type ClampedLength<
+  T extends IterableContainer,
+  Offset extends number,
+  Length extends number,
+> =
+  IsNegative<Length> extends true
+    ? 0
+    : number extends T["length"]
+      ? number
+      : GreaterThan<Sum<Offset, Length>, T["length"]> extends true
+        ? IfLiteral<Subtract<T["length"], Offset>, 0>
+        : Length;
+
+type FixedLengthSplice<
+  T extends IterableContainer,
+  Start extends number,
+  DeleteCount extends number,
+  Replacement extends IterableContainer,
+  PositiveStart extends number = PositiveIndex<T, Start>,
+> = PositiveStart extends unknown
+  ? ArraySplice<
+      T,
+      PositiveStart,
+      ClampedLength<T, PositiveStart, DeleteCount>,
+      Replacement
+    >
+  : never;
+
+/** Widen a tuple into an array type. */
+type WidenTuple<T extends IterableContainer> =
+  T extends ReadonlyArray<infer Item>
+    ? IfNever<Item, ReadonlyArray<unknown>, Array<WidenLiteral<Item>>>
```

Comment:
I see you wrap WidenTuple everywhere with Readonly<>, why not define it as readonly here?

```suggestion
    ? IfNever<Item, ReadonlyArray<unknown>, ReadonlyArray<WidenLiteral<Item>>>
```